go/src/url_shortener/url_shortener.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+)
+
+var (
+	cache URLShortener = URLShortener{shorts: make(map[string]*urlInfo)}
+)
+
+// URLShortener URL shortener server data structure
+type URLShortener struct {
+	mux    sync.Mutex
+	shorts map[string]*urlInfo
+}
+
+type urlInfo struct {
+	longURL string
+
+	// statistics
+	shortURL string
+	//count    int
+}
+
+func (cache *URLShortener) addURL(longURL string, shortURL string) {
+	cache.mux.Lock()
+	defer cache.mux.Unlock()
+
+	cache.shorts[shortURL] = &urlInfo{longURL, shortURL}
+}
+
+func (cache *URLShortener) getURL(shortURL string) (string, error) {
+	cache.mux.Lock()
+	defer cache.mux.Unlock()
+
+	if info, ok := cache.shorts[shortURL]; ok {
+		return info.longURL, nil
+	}
+
+	return "", fmt.Errorf("short URL not found: %s", shortURL)
+}
+
+func shorten(longURL string) string {
+	hasher := sha1.New()
+
+	hasher.Write([]byte(longURL))
+	sum := hasher.Sum(nil)
+
+	shortURL := fmt.Sprintf("%x", sum)[:6]
+
+	return shortURL
+}
+
+func shortenerHandler(w http.ResponseWriter, r *http.Request) {
+	longURL := r.URL.Path[len("/shorten/"):]
+	shortURL := shorten(longURL)
+
+	cache.addURL(longURL, shortURL)
+
+	fmt.Fprintf(w, "/%s", shortURL)
+}
+
+func statsHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Some statistics")
+}
+
+func expanderHandler(w http.ResponseWriter, r *http.Request) {
+	shortURLCandidate := r.URL.Path[len("/"):]
+
+	redirectURL, err := cache.getURL(shortURLCandidate)
+
+	if err != nil {
+		fmt.Fprintf(w, "%s\n", err)
+	}
+
+	http.Redirect(w, r, redirectURL, http.StatusSeeOther)
```

Comment:
if the short URL is not known, here we print an error but still performs the HTTP redirect since there's no `return` in the `if`

Reply:
you may want to return a different status code in this case. StatusNotfound maybe?

Reply:
Thank you! I fix that and add some statistics.

Reply:
How do I return a status code? Is there an API to do that (I do not remember).

Reply:
Look at WriteHeader https://golang.org/pkg/net/http/#ResponseWriter.